Add `napari_experimental_provide_theme` hook specification

lukasz-migas opened the pull request:
This PR is an offshot of #2900 that will enable easy addition of icons, stylesheets and themes. Since there is still a bit more to do there, the `napari_provide_theme` spec was separated as it already provides the relevant functionality.

# Description
- Add a way to provide `theme` that controls the look of napari
- Add new `Theme` class that is an evented model
- Changes to the active theme are now immediately visible in the app

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# References
#2900 - original PR
#2994  - evented dict

# How has this been tested?
- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

![evented-theme](https://user-images.githubusercontent.com/25161821/125358584-60cc3900-e369-11eb-957d-b07140e6e1c1.gif)

lukasz-migas committed:
Add hook specification to add new themes, icons and stylesheets

- added new hook specification to allow addition of new themes, icons and stylesheets to napari
- minor improvements to how qt resources are loaded

lukasz-migas committed:
Update docstrings

- updated docstrings

lukasz-migas committed:
Merge branch 'napari:master' into theme-hook

lukasz-migas committed:
WIP: added evented dict container, split hook spec into three hooks

- split the hook implementation into three separate hooks (qss, icons and theme)
- added new container class: `EventedDict` and `TypedMutableMapping` so we can migrate themes to an evented class
- refactored how theme plugin data is read

lukasz-migas committed:
Theme dict is now evented, immediate updates to stylesheet

- theme dict is now an evented model
- changes to any theme data will result in immediate update of the stylesheet
- minor bug fix to the evented dict not propagating child events

lukasz-migas committed:
Merge branch 'napari:master' into theme-hook

lukasz-migas committed:
Themes now reflect state of plugin, preferences shows all available themes

- themes are now dynamically updated whenever a plugin is enabled/disabled/removed
- theme dropdown is now updated whenever new theme is added/removed
- plugin manager now displays more useful information if theme data failed validation

lukasz-migas committed:
Merge branch 'theme-hook' of https://github.com/lukasz-migas/napari into theme-hook

lukasz-migas committed:
Icon colors can be changed dynamically

- fixed issue in settings tests (the test should have always failed since it was passing wrong data to the `register_theme` function)
- found a workaround for dynamically updating icon colors without causing crash...

lukasz-migas committed:
Tidy up

- removed unused code

lukasz-migas committed:
[WIP] Add pydantic-based Theme, refactor theme

- added pydantic-based Theme class with better validation of values
- modified the way `get_theme` is retrieved. Currently it will trigger warnings whenever somebody calls it which might not be desired
- removed the `palettes` warning from utils/theme.py since it was meant to be removed in 0.4.6

lukasz-migas committed:
Update qt_main_window.py

- existing themes will now be properly connected at startup

lukasz-migas committed:
Added theme tests, theme improvements

- removed custom Themes model since its not necessary
- added a couple extra tests to handle new theme model (wip)
- improved docs regarding new theme hooks

lukasz-migas committed:
Temporarily remove evented dict

lukasz-migas committed:
Merge branch 'napari:master' into theme-hook

lukasz-migas committed:
Minor improvements

- changes to canvas/console and syntax style will now immediately update the ui
- minor changes to Theme model

lukasz-migas committed:
Disabled warning

lukasz-migas committed:
Update vispy_canvas.py

lukasz-migas committed:
Minor bug fixes

Fixes to `UserWarning` warnings when transforming color to 0-1 scale

lukasz-migas committed:
Remove icons/stylesheet hooks

- temporarily remove stylesheet/icon hooks
- rename `napari_experimenta_provide_theme` to `napari_provide_theme` hook spec
- auto-format few pesky files

lukasz-migas committed:
Merge branch 'napari:master' into hook-theme-only

lukasz-migas committed:
Tidy up

- remove unused files
- minor bug fixes and typos

lukasz-migas committed:
Update docs

sofroniewn commented on examples/new_theme.py:
```diff
@@ -13,20 +14,23 @@
+blue_theme.icon = (
+    'rgb(0, 255, 255)'  # you can provide colors as rgb(XXX, YYY, ZZZ)
 )
+blue_theme.background = 28, 31, 48  # or as tuples
+blue_theme.foreground = [45, 52, 71]  # or as list
+blue_theme.primary = '#50586c'  # or as hexes
+blue_theme.current = 'orange'  # or as color name
```
Love this!

sofroniewn commented on napari/utils/theme.py:
```diff
@@ -13,64 +15,95 @@
+]
+
+
+class Theme(EventedModel):
```
Very nice

sofroniewn reviewed (commented):
This looks great! I think we just need to fix the failing tests - looks like there might still be some event connection left over that triggers an update to a deleted object - see 

```python-traceback
==================================== ERRORS ====================================
  ___________________ ERROR at teardown of test_changing_theme ___________________
  
  qtbot = <pytestqt.qtbot.QtBot object at 0x7ff551c75cd0>
  request = <SubRequest 'make_napari_viewer' for <Function test_changing_theme>>
  napari_plugin_manager = <napari.plugins._plugin_manager.NapariPluginManager object at 0x7ff551c75ee0>
  
      @pytest.fixture
      def make_napari_viewer(
          qtbot, request: 'FixtureRequest', napari_plugin_manager
      ):
          """A fixture function that creates a napari viewer for use in testing.
      
          Use this fixture as a function in your tests:
      
              viewer = make_napari_viewer()
      
          It accepts all the same arguments as napari.Viewer, plus the following
          test-related paramaters:
      
          ViewerClass : Type[napari.Viewer], optional
              Override the viewer class being used.  By default, will
              use napari.viewer.Viewer
          strict_qt : bool or str, optional
              If True, a check will be performed after test cleanup to make sure that
              no top level widgets were created and *not* cleaned up during the
              test.  If the string "raise" is provided, an AssertionError will be
              raised.  Otherwise a warning is emitted.
              By default, this is False unless the test is being performed within
              the napari package.
              This can be made globally true by setting the 'NAPARI_STRICT_QT'
              environment variable.
          block_plugin_discovery : bool, optional
              Block discovery of non-builtin plugins.  Note: plugins can still be
              manually registered by using the 'napari_plugin_manager' fixture and
              the `napari_plugin_manager.register()` method. By default, True.
      
          Examples
          --------
          >>> def test_adding_shapes(make_napari_viewer):
          ...     viewer = make_napari_viewer()
          ...     viewer.add_shapes()
          ...     assert len(viewer.layers) == 1
      
          >>> def test_something_with_plugins(make_napari_viewer):
          ...     viewer = make_napari_viewer(block_plugin_discovery=False)
      
          >>> def test_something_with_strict_qt_tests(make_napari_viewer):
  napari/utils/events/evented_model.py:155: in __setattr__
      getattr(self.events, name)(value=after)  # emit event
  napari/utils/events/event.py:575: in __call__
      self._invoke_callback(cb, event)
  napari/utils/events/event.py:598: in _invoke_callback
      _handle_exception(
  napari/utils/events/event.py:596: in _invoke_callback
      cb(event)
  napari/_qt/qt_main_window.py:1561: in _update_theme
      self.qt_viewer.viewer.theme = value
  napari/components/viewer_model.py:323: in __setattr__
      return super().__setattr__(name, value)
  napari/utils/events/evented_model.py:155: in __setattr__
      getattr(self.events, name)(value=after)  # emit event
  napari/utils/events/event.py:575: in __call__
      self._invoke_callback(cb, event)
  napari/utils/events/event.py:598: in _invoke_callback
      _handle_exception(
  napari/utils/events/event.py:596: in _invoke_callback
      cb(event)
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
  
  self = <[RuntimeError('Internal C++ object (QtAboutKeyBindings) already deleted.') raised in repr()] QtAboutKeyBindings object at 0x7ff5843c5fc0>
  event = <Event blocked=False handled=False native=None source=None sources=[] type='theme'>
  
      def update_active_layer(self, event=None):
          """Update the active layer and display key bindings for that layer type.
      
          Parameters
          ----------
          event : napari.utils.event.Event, optional
              The napari event that triggered this method, by default None.
          """
          theme = get_theme(self.viewer.theme)
          col = theme['secondary']
      
          # Add class and instance viewer key bindings
          active_shortcuts = action_manager._get_active_shortcuts(
              self.key_map_handler.active_keymap
          )
          text = get_key_bindings_summary(active_shortcuts, col=col)
      
          # Update layer speficic key bindings if all active are displayed
          self.key_bindings_strs[self.ALL_ACTIVE_KEYBINDINGS] = text
  >       if self.layerTypeComboBox.currentText() == self.ALL_ACTIVE_KEYBINDINGS:
  E       RuntimeError: Internal C++ object (PySide2.QtWidgets.QComboBox) already deleted.
  
  napari/_qt/dialogs/qt_about_key_bindings.py:151: RuntimeError
  =================================== FAILURES ===================================
```

lukasz-migas committed:
Disconnect event

- explicitly disconnect the theme event in `qt_about_key_bindings` dialog

lukasz-migas committed:
Update test_qt_about_key_bindings.py

tlambert03 commented on napari/utils/theme.py:
```diff
@@ -13,64 +15,95 @@
+class Theme(EventedModel):
+    """Theme model."""
+
+    folder: str
```
`folder` seems pretty antiquated at this point.  This started out as the literal folder that the icons were in, but we generally use it as the name of the theme right?

lukasz-migas replied:
I agree. The `folder` attribute could be renamed to `name`. Now would be a good time to do it since we are adding a completely new model to handle theme data.

sofroniewn replied:
Let's rename to "name" in this PR. But should we then reuse this "name" inside our dict of themes, i.e. should this now be the name you use when registering and getting a theme too, or can they be different?

tlambert03 commented on napari/utils/theme.py:
```diff
@@ -13,64 +15,95 @@
+    """Theme model."""
+
+    folder: str
+    syntax_style: str
```
If syntax style must be one of the supported themes from pygments, then we should either use a string validator here or make it an enum of acceptable values.

lukasz-migas replied:
I've added a value validator that will check against `pygments.styles.STYLE_MAP`. I don't think `pygments` is a direct dependency so this might have to optional?

tlambert03 replied:
> I don't think pygments is a direct dependency so this might have to optional?

That's true, but the only reason we have this option in the theme is basically to support `napari-console`, which depends on qtconsole, and _that_ depends on `pygments` (and if it stopped depending on pygments, then we'd basically need to come up with new themeing anyway).  So, I'd say we should either remove this from our theme entirely (and leave it to `napari-console` to figure out how to theme itself)... or we should just use pygments for now and leave that decision for when we don't require `napari-console`.  If we keep it, here's one option for validating against a set of string literals:

```py
from typing_extensions import Literal as L
import pygments

class Theme(EventedModel):
    s: Union[tuple(L[x] for x in pygments.styles.STYLE_MAP)]
```

sofroniewn replied:
> or we should just use pygments for now and leave that decision for when we don't require napari-console

Maybe we do this for now and then refactor when we pull out napari-console? Just seems a little easier, but we can also deal with now

lukasz-migas replied:
Agree, I am happy to change this when napari-console is no longer a dependency in the future.

tlambert03 commented on napari/plugins/hook_specifications.py:
```diff
@@ -496,3 +496,29 @@ def napari_experimental_provide_dock_widget() -> Union[
+    """Provide GUI with a set of colors used through napari. This hook allows you to
+    provide additional color schemes so you can accomplish your desired styling.
+
+    Themes are provided as `dict` with several required fields and correctly formatted
```
since `folder` is just an alias for `name`, we probably don't need the additional dict here, right? (seems like the key is always just the `folder` string... which we can call `name`.  then this can just be `Sequence[Dict[str, Union[str, Tuple, List]]]`

lukasz-migas replied:
Thanks for poinint this out. I've changed to return a sequence of dictionaries.

tlambert03 commented on napari/plugins/hook_specifications.py:
```diff
@@ -496,3 +496,29 @@ def napari_experimental_provide_dock_widget() -> Union[
+
+    Examples
+    --------
+    """
```
Please add example here as well since this goes in the docs on the webpage... 
Also please add a new `.. autofunction:: napari_provide_theme` entry somewhere in `docs/plugins/hook_specifications.rst` (That's still not done automatically)

tlambert03 commented on napari/utils/theme.py:
```diff
@@ -156,11 +198,31 @@ def get_theme(name):
-        return theme.copy()
+        _theme = theme.copy()
+        if as_dict:
+            # warnings.warn(
```
I'm for this.... will we uncomment?

lukasz-migas replied:
For the time being, I've left it commented out since a lot of tests will fail with this warning.

sofroniewn replied:
Agreed we should probably uncomment. We could add to our test skips here https://github.com/napari/napari/blob/115d3ef468d4babb8d722513a1f97002811394a8/pyproject.toml#L84-L96 temporarily and then fix tests in follow up PRs, or fix tests still in this PR

lukasz-migas replied:
Done 👍

tlambert03 commented on examples/new_theme.py:
```diff
@@ -5,6 +5,7 @@
 from skimage import data
 import napari
 from napari.utils.theme import available_themes, get_theme, register_theme
+from napari._qt.qt_resources._icons import register_napari_theme
```
will registering a new theme always mean calling both `register_theme()`, followed by `register_napari_theme()`?  Can we call `register_napari_theme` internally whenever a new theme has been registered?
If not, and this needs to be public, it should be exposed in `napari.qt`

sofroniewn replied:
Agreed, it would be nice if we could call `register_napari_theme` internally and this did not need to be exposed publicly

lukasz-migas replied:
This is now done internally whenever a new theme is registered in the `_theme` dict so the user doesn't need to call it.

tlambert03 commented on examples/plugins/theme_plugin.py:
```diff
@@ -0,0 +1,30 @@
+from napari_plugin_engine import napari_hook_implementation
+
+
+def custom_theme():
```
this whole example can probably just go in the docstring of the hook_specification, where it will show up on the web docs

sofroniewn replied:
I'm ok with having it in both places, I sort of like the runnable example in a plugins folder, but example should be in doc string too probably

tlambert03 replied:
but, it's not exactly "runnable" that easily.  It _can_ be runnable, but only if someone uses `napari  examples/plugins/theme_plugin.py` (but just running `python examples/...` would do nothing).  That would be a first for our examples directory.  So at the very least, it should be made clear that you can't actually run this example unless you start it "just so"

sofroniewn replied:
ok good point, maybe we should just move to doc string for now

tlambert03 commented on napari/utils/settings/_defaults.py:
```diff
@@ -282,6 +281,15 @@ class NapariConfig:
         # Napari specific configuration
         preferences_exclude = ['schema_version']
 
+    def refresh_themes(self):
```
This can come in another PR, but I wonder whether we would have designed the `Appearance.theme` field as a `str` here if we had a pydantic model in the first place.  Now it feels kinda funny to have a `class Theme(str)` above instead of just `theme: utils.theme.Theme`  @sofroniewn ?

sofroniewn replied:
Would this then allow someone in the settings to immediately create custom themes where they changed individual colors etc. That would be quite nice, but let's leave it for its own PR

tlambert03 commented on napari/plugins/_plugin_manager.py:
```diff
@@ -289,6 +291,75 @@ def available_samples(self) -> Tuple[Tuple[str, str], ...]:
+                    )
+                    warn(message=warn_message)
+                    continue
+                _data[name] = theme_colors
```
Can you comment on why we're storing the theme data instead of theme instances?
Also, it kinda feels like we're duplicating the validation that Pydantic already does for us here.  how about

```py
try:
    theme = Theme.pars_obj(data)
except Exception as e:  # or pydantic.ValidationError
    warn(f'skipping bad theme: {data}')
```

lukasz-migas replied:
You are right, I've changed this so `pydantic` does all of the necessary validation.

tlambert03 commented on napari/utils/theme.py:
```diff
@@ -13,64 +15,95 @@
+from .events import EventedModel
+from .events.containers._evented_dict import EventedDict
+
+THEME_KEYS = [
```
not needed, pydantic provides `list(Theme.__fields__)`.  But also, see my comment down in plugin manager, where this seems to be used.  I would just let pydantic do the validation for us (that's the big win of using a pydantic model for the theme).  No need to check the keys and values and all that stuff ourself

tlambert03 commented on napari/_qt/qt_main_window.py:
```diff
@@ -440,6 +452,131 @@ def __init__(self, viewer, *, show: bool = True):
+        # clear out all data to make sure it's up to date (e.g. after
+        # plugin was disabled or unregistered)
+        plugin_manager._theme_data.clear()
+        plugin_manager.discover_themes()
```
discover_themes() calls `hook.napari_provide_theme.call_historic` (which only needs to be called once) but this is connected to an event that will trigger this every time a new plugin is registered... is the historic hook still necessary?  (a historic hook is usually only ever called once, and you provide a callback here, `register_theme_colors` that will be called _for you_ each time the plugin manager detects a new plugin that provides the hook)

lukasz-migas replied:
This was actually not necessary so I've removed it. Thanks!

tlambert03 commented on napari/_qt/qt_main_window.py:
```diff
@@ -385,6 +386,10 @@ def __init__(self, viewer, *, show: bool = True):
         self._add_help_menu()
 
+        # discover any themes provided by plugins
+        plugin_manager.discover_themes()
```
other than triggering discovery when we show the first window, it's not immediately clear to me why all of this logic now lives on the mainwindow.    I feel like we're mixing business logic from the plugin_manager and the settings altogether on the main window here.  most of the new theme-related methods aren't really using the `self` attribute (except to call `_rebuild_theme_settings`, which is just a static method that perhaps belongs on the settings manager instead?)

lukasz-migas replied:
I've removed a couple of the methods since they were not really required. 

The remaining methods such as `_setup_existing_themes`, `_add_theme`, `_remove_theme`, `_connect_theme` and `_theme_icon_changed` methods are added in the `MainWindow` so any changes to the theme values are reflected in the GUI.

tlambert03 commented on napari/_qt/qt_main_window.py:
```diff
@@ -440,6 +452,131 @@ def __init__(self, viewer, *, show: bool = True):
+    def _connect_theme(self, theme):
+        # connect events to update theme. Here, we don't want to pass the event
+        # since it won't have the right `value` attribute.
+        theme.events.background.connect(lambda _: self._update_theme())
```
fyi, I believe you can connect directly to `theme.events.connect` so :`theme.events.connect(lambda _: self._update_theme())` gets them all done in one go  (or at least I think 🤔)

lukasz-migas replied:
That's true. I did think of doing it like this and we still can. I've decided against it because the `icon` attribute is a special case where the `_update_theme` method would do nothing.

tlambert03 reviewed (commented):
Pretty awesome stuff @lukasz-migas!

I left a lot of comments... hope you don't mind.

a couple general thoughts that don't really go anywhere specific: 

I think it's pretty cool that we can dynamically change themes now... but do you see this eventually being something that persists across sessions?  If so, it feels like "dark theme" doesn't really mean much anymore, since it is whatever you make it.  If not, it feels significantly less powerful.  So perhaps we need the provided themes to be frozen, and once you start mucking about, you're creating a "custom" theme, but not actually changing the frozen themes in `_themes`

this can come in a follow up PR, but I feel like we're not using the Pydantic model to it's fullest.  With this we now have even more "theme aware" parts.  The plugin manager gains a few themes, settings have some new theme stuff.  The main window has a lot more theme methods... so I generally feel like it could be a little better compartmentalized. 

anyway, let me know if you have thoughts/questions on my comments

lukasz-migas committed:
Tidied up, removed unnecessary methods, simplified behaviour

- Add examples to the hook specification
- Removed unnecessary methods
- Added console `style_syntax` validation method
- Changed `napari_provide_theme` signature
- Added more tests

lukasz-migas committed:
Bring up to date

lukasz-migas committed:
Improved testing and theme validation

- the `_theme_data` dict on plugin manager now stores the `Theme` object
- improved theme validation
- improved tests

lukasz-migas committed:
Bring up to date

lukasz-migas committed:
Changed qt resources hash calculation

- changed the way qt resources hash is computed
- fixed potential future bug where the `os.path` could be overwritten by variable named `path`

sofroniewn commented on napari/_vispy/vendored/volume.py:
```diff
@@ -530,8 +530,8 @@
 
 
 class VolumeVisual(Visual):
```
Can you removed the changed to this vendored file. It will likely cause merge conflicts with some other PRs in progress and they are only formatting changes

sofroniewn commented on napari/utils/misc.py:
```diff
@@ -463,23 +465,44 @@ def dir_hash(path: Union[str, Path], include_paths=True, ignore_hidden=True):
+    include_paths: bool = True,
+    ignore_hidden: bool = True,
+) -> str:
+    """Compute the hash of list of paths."""
```
Can you add a "parameters" section to the doc string here

sofroniewn commented on napari/utils/misc.py:
```diff
@@ -463,23 +465,44 @@ def dir_hash(path: Union[str, Path], include_paths=True, ignore_hidden=True):
     return _hash.hexdigest()
 
 
+def _file_hash(_hash, file: Path, path: Path, include_paths: bool = True):
```
Can you add a "parameters" section to the doc string here

lukasz-migas replied:
Done

sofroniewn commented on napari/utils/misc.py:
```diff
@@ -448,12 +450,12 @@ def pick_equality_operator(obj) -> Callable[[Any, Any], bool]:
 
 
-def dir_hash(path: Union[str, Path], include_paths=True, ignore_hidden=True):
+def dir_hash(
```
Can you add a "parameters" section to the doc string here. (I know it was already there without one, but now is a good time to add it, thanks!!)

sofroniewn commented on napari/utils/theme.py:
```diff
@@ -13,64 +16,117 @@
+    {
+        'dark': Theme(
+            **{
+                'folder': 'dark',
```
See comment above - but if we rename `folder` to `name`, is it weird if we then have a them with a different name in the `EventedDict` and in the `Theme` itself? Might be nice to force these to be the same somehow. Thoughts @tlambert03?

lukasz-migas replied:
I actually torn on this because I can envisage a situation where a theme is just a variation of an existing theme e.g. `dark` so the icon color is the same and there should be no need to create a new version of the icons under a different  namespace.

sofroniewn replied:
ok, that's fine

sofroniewn reviewed (commented):
Looks like this is coming along nicely @lukasz-migas - I just did another review, flagged some questions, mostly minor. Hopefully we should be able to get this in soon!!

lukasz-migas committed:
Added more tests, few bug fixes and general improvements

- renamed `folder` to `name` in the Theme model
- renamed `folder` to `name` in the stylesheet templates
- disabled memory hungry test
- added more tests for the plugin hookspec behaviour
- removed unnecessary methods from the MainWindow
- removed unused code
- added docs

lukasz-migas committed:
Brought up to date with master

lukasz-migas committed:
Merge branch 'master' into hook-theme-only

lukasz-migas committed:
Added more tests

- removed unused method
- added more tests

lukasz-migas committed:
Fixed tests

lukasz-migas committed:
Bug fix?

tlambert03 commented on examples/cursor_ray.py:
```diff
@@ -11,18 +11,21 @@
-        [1, 1, 1],
-    ]
-) * sidelength_data
+bounding_box = (
```
my black also corrects these... but this bothers my sense of PR granularity :joy:  (i.e. when we look back to see what PR modified a file, if we're going to do a pure formatting change on an unrelated file, I'd like the PR to be something like "black reformat", not "add hook spec")

can you revert this file, and other un-related file formats (like ndLabels... not sure if there are others)?

do you have a `--all-files` setting in your `pre-commit` or black settings somehow?  I'm curious why this would have been modified at all?

lukasz-migas replied:
I am using the pre-commit config without any modifications. I will check which files were just formatted and undo them.

tlambert03 reviewed (commented):
@lukasz-migas 

can you undo all of the formatting-only changes that don't have anything to do with the context of this PR?  makes it a bit harder to review...

lukasz-migas committed:
Undo auto-formatting

lukasz-migas committed:
Undo auto-formatting

tlambert03 commented on napari/utils/theme.py:
```diff
@@ -156,11 +181,27 @@ def get_theme(name):
-        return theme.copy()
+        _theme = theme.copy()
+        if as_dict:
+            warnings.warn(
```
quick note here, with this warning, everyone will be seeing a warning on every napari usage now because of napari-console:
```
/Users/talley/miniconda3/envs/napdev/lib/python3.9/site-packages/napari_console/qt_console.py:161: FutureWarning: Themes were changed to use evented model with Pydantic's color type rather than the `rgb(x, y, z)`. You can get the old color by calling `color.as_rgb()`. The `as_dict=True` option will be removed in 0.X.X
  theme = get_theme(self.viewer.theme)
```

lukasz-migas replied:
That's true - its can be very easily fixed in the `napari-console` plugin but until then it will be showing each time. 

The startup warning is actually caused by the fact that I am connecting the `theme.console` and `theme.syntax_style` events directly to the `qt_viewer.console._update_theme` methods. I could wrap with `warnings.catch_warnings` for the time being?

tlambert03 replied:
I do think it's pretty jarring to have everyone see a warning in our default install.  So either here, or in a follow-up PR before we release, we should either do the catch_warnings... or we should update napari-console and pin our dependency here

lukasz-migas committed:
Renamed hook spec, fixed warning, removed example

- removed the `theme_plugin` example implementation - one can be found in the docs
- renamed `napari_provide_theme` to `napari_experimental_provide_theme`
- added catch warning to prevent qtconsole warnings at the start

lukasz-migas committed:
Bring up to date

lukasz-migas committed:
Merge branch 'master' into hook-theme-only

lukasz-migas committed:
Fix failed test

lukasz-migas committed:
Merge branch 'napari:master' into hook-theme-only

lukasz-migas committed:
Merge branch 'napari:master' into hook-theme-only

lukasz-migas committed:
Temporarily disable few tests...

lukasz-migas committed:
Merge branch 'napari:master' into hook-theme-only

lukasz-migas committed:
Disable more tests

lukasz-migas committed:
Undo disable tests

lukasz-migas committed:
Bring up to date

lukasz-migas committed:
Merge branch 'master' into hook-theme-only

lukasz-migas committed:
Update qt_main_window.py

lukasz-migas committed:
Uncomment all tests

lukasz-migas committed:
Update 02_custom.qss

lukasz-migas committed:
Update qt_main_window.py

sofroniewn reviewed (approved):
They are all passing!!

Heroic work here and well figured out with the teardown. I'd like to merge this now so that it gets into the next release candidate, but I'll wait for @tlambert03 to approve too.

Thanks again, and onwards and upwards! 🎆 🚀
